### ISSUE TITLE
Relative Pixel RMSE Metrics obtained with new simulated dataset are larger than expected

### ISSUE THREAD

**Issue by jeipollack** [CosmoStat/wf-psf#169]
A new script was used to generate a validation dataset that should include the same features as the dataset used in the original WaveDiff paper, which used the `coherent_euclid_dataset`.  The difference would be in the random seed that assigns SEDs to stars and pixel level noise.

The relative pixel RMSE metrics obtained when WaveDiff `main` and newer versions are larger than those reported in the paper.

Tobias asked to do a full run on the dataset used for the paper to see if we can reproduce the results of the paper, which would mean it's just by chance that we ended up with degraded metrics.  Else, there is a bug somewhere.

**Comment by jeipollack** [comment]
Note-to-self:  Launched `main` branch runs in `/lustre/fswork/projects/rech/ynx/uuu68hq/validation/dataset/main/coherent_euclid_dataset` with:
* paper's training hyper parameters 
```
      # Learning rates for the parametric parts. It should be a str where numeric values are separated by spaces.
      learning_rate_params: [1.0e-2, 0.4e-2]

      # Learning rates for the non-parametric parts. It should be a str where numeric values are separated by spaces."
      learning_rate_non_params: [1.0e-1, 6.0e-1]

      # Number of training epochs of the parametric parts. It should be a strign where numeric values are separated by spaces."
      n_epochs_params: [15, 15]

      # Number of training epochs of the non-parametric parts. It should be a str where numeric values are separated by spaces."
      n_epochs_non_params: [100, 50]
```
* Current training hyper parameters 
```
      # Learning rates for the parametric parts. It should be a str where numeric values are separated by spaces.
      learning_rate_params: [1.0e-2, 1.0e-2]

      # Learning rates for the non-parametric parts. It should be a str where numeric values are separated by spaces."
      learning_rate_non_params: [1.0e-1, 1.0e-1]

      # Number of training epochs of the parametric parts. It should be a strign where numeric values are separated by spaces."
      n_epochs_params: [20, 20]

      # Number of training epochs of the non-parametric parts. It should be a str where numeric values are separated by spaces."
      n_epochs_non_params: [100, 120]
```

**Comment by tobias-liaudat** [comment]
Thank you very much for taking a look at this @jeipollack , it is very important.

I'll just add some info. I ran the model from the branch [feature/real_data_metrics_lazyloading](https://github.com/CosmoStat/wf-psf/tree/feature/real_data_metrics_lazyloading) with the original dataset (from the paper), which I named v.0.0.0,  and then ran the model from an older branch [u/tobias-liaudat/feature-phase_retrieval](https://github.com/CosmoStat/wf-psf/tree/u/tobias-liaudat/feature-phase_retrieval), which was perfectly matching results from the release [v1.3.0](https://github.com/tobias-liaudat/wf-psf/tree/v1.3.0) of the v1 version of the code.

With the `u/tobias-liaudat/feature-phase_retrieval`, I got around a `0.9%` relative pixel rmse error at observation resolution in the test dataset. This is consistent with what we got in the paper that is `0.86%`.

With the current `feature/real_data_metrics_lazyloading`, which is very similar to the branches with new features (but adding the chi2 metric), I got around `4%` for the same metric.

The metrics were run using the checkpoint model, as usual.

This seems to suggest we have a hidden bug somewhere. 

For info, the jobs, configs and results are in the `wf-psf-validation` repo and in JZ can be found here `/lustre/fswork/projects/rech/ynx/commun/simulated_datasets_v2`

**Comment by jeipollack** [comment]
Hi @tobias-liaudat, I see in your logs what I noted yesterday investigating the checkpoint issue #168 .  In the second cycle of training, we get stuck in a local minima when the second cycle of training starts for both parametric and non-parametric parts.

For the `poly` model we're monitoring `mean_squared_error`, if we change to monitoring the `loss` as was done for `v0_2_0` (unitary masks) the results match `main` and the best weights are saved at the same epoch 116 in cycle 2.

I am running training on different commit states to see if I can spot what leads to change and looking carefully at the before and earlier versions to see if I can spot the mistake.

**Comment by tobias-liaudat** [comment]
@jeipollack thanks for investigating this.

One thing that could be useful to test is if, instead of comparing the metrics with the `checkpoint` model, we do that with the `final` model. We should see if the results match when using the original dataset (from the paper) between the new and the older versions of the model. Then we could narrow down the problem to the checkpointing of the model and not the training.

**Comment by jeipollack** [comment]
@tobias-liaudat I noticed though with the `final model` we're saving the weights of the last epoch not the weights from the best epoch.  That may be the reason why the metrics of `main` and newer branches like `gh-165` agreed.

**Comment by jeipollack** [comment]
I confirmed that metrics evaluators in branch PR #166 (issue #165) reproduce those of the `main_random_seed` branch which updates tag 1.20 with a random seed.  I would also assume based on this that the issue is not the PSF model at least when used for inference/metrics evaluation.

So, the reason for the discrepant results has to do solely with training.  But, I am waiting to confirm that the training of the `main` branch (tag: 2.0.1). I had to restart the training (yesterday evening) after noting that the learning rate for the non-parametric part was not changed to the value used in `main_random_seed`.  

I know it was suggested to look at a forked repo, but I prefer to first verify the tagged versions of WaveDiff in the official repo. 
Then I will properly use the `git bisect` method to identify the commit which changed things.  For this, the commit id of the merged forked branch could be used as a starting point before the regression happened.

**Comment by jeipollack** [comment]
I have updated the [table](https://docs.google.com/spreadsheets/d/1jIyvXOCuHDY8P548oaWMtbU2hrLgxuKEvgL8D92dNOU/edit?gid=0#gid=0) with the results of running branches `main`, `case_study`, `gh-165 (inference + PSF model refactor + Zernike contribution enhancement)` on the v0xx and v1xx datasets.

I also launched runs for different configurations using the training hyper parameters for SHE as well as those for WaveDiff paper for 2k stars, and by mistake a tweaked version of the paper for 2k stars that actually reduced the Pixel ABS & Relative RMSE Errors.  I decided to submit new runs for `v0xx` and `v1xx` for consistency checks.  This includes tweaking the first epoch parametric learning rate for the SHE training hyper parameters just to see if there is an impact.

| SHE                                |                      |
| -----------------------| ------------ |
| learning_rate_params  | [0.01, 0.01] |
| learning_rate_nonparams | [0.1, 0.1] |
| n_epochs_params     | [20, 20] | 
| n_epochs_nonparams     | [100, 120] | 

| WaveDiff Paper    2k stars    |                      |
| -----------------------| ------------ |
| learning_rate_params  | [0.01, 0.004] |
| learning_rate_nonparams | [0.1, 0.06] |
| n_epochs_params.       | [15, 15] | 
| n_epochs_nonparams  | [100, 50] | 

| Tweaked WaveDiff Paper    2k stars    |                      |
| -----------------------| ------------ |
| learning_rate_params  | [**0.1**, 0.004] |
| learning_rate_nonparams | [0.1, 0.06] |
| n_epochs_params.       | [15, 15] | 
| n_epochs_nonparams  | [100, 50] | 

| Tweaked SHE       |                      |
| -----------------------| ------------ |
| learning_rate_params  | [**0.1**, 0.004] |
| learning_rate_nonparams | [0.1, 0.06] |
| n_epochs_params.       | [20, 20] | 
| n_epochs_nonparams  | [100, 120] |


You can look specifically at the tabs `Test Pixel & OPD metrics` and `Test Shape Metrics`.

There is something strange with the `v1_3_0` dataset which in comparison to `v1_1_0` and `v1_2_0` should only apply a realistic mask. The metrics are poor. Unfortunately, the monolithic script added to the PR with the multi-nested conditionals is difficult to follow. It contains no tests; nor details to how it was validated; there are hard-coded parameters. Since it doesn't meet the code quality standards of WaveDiff v2.0+ I am considering pulling the commits out into a separate branch for generation of validation datasets.
